components/analytics-data-connectors/org.wso2.carbon.analytics.datasource.mongo/src/main/java/org/wso2/carbon/analytics/datasource/mongo/MongoAnalyticsRecordStore.java

```diff
@@ -0,0 +1,265 @@
+        if (recordGroup.getTimeFrom() == Long.MIN_VALUE && recordGroup.getTimeTo() == Long.MAX_VALUE) {
+            iterableResult = db.getCollection(collection).find().projection(fields(include("_id", "timestamp", "data")));
+            if (count != -1) {
+                iterableResult = iterableResult.limit(count);
+            }
+        } else {
+            iterableResult = db.getCollection(collection).find()
+                    .filter(and(gte("timestamp", recordGroup.getTimeFrom()), lt("timestamp", recordGroup.getTimeTo())));
+            if (count != -1) {
+                iterableResult = iterableResult.limit(count);
+            }
+        }
+        return new MongoRecordDataIterator(iterableResult.iterator(), columns, tableName, tenantId);
+    }
+
+    private AnalyticsIterator<Record> readRecordsByIds(GlobalMongoRecordGroup recordGroup) throws AnalyticsException {
+        return this.lookupRecordsByIds(recordGroup.getTenantId(), recordGroup.getTableName(), recordGroup.getIds(),
+                recordGroup.getColumns());
+    }
+
+    private AnalyticsIterator<Record> lookupRecordsByIds(int tenantId, String tableName, List<String> ids, List<String> columns) {
+        String collection = this.generateTargetCollectionName(tenantId, tableName);
+        MongoCursor<Document> cursor = db.getCollection(collection).find(in("_id", ids)).iterator();
+        return new MongoRecordDataIterator(cursor, columns, tableName, tenantId);
+    }
+
+    @Override
+    public long getRecordCount(int tenantId, String tableName, long timeFrom, long timeTo)
+            throws AnalyticsException, AnalyticsTableNotAvailableException {
+        String collection = this.generateTargetCollectionName(tenantId, tableName);
+        return db.getCollection(collection).count(and(gte("timestamp", timeFrom), lt("timestamp", timeTo)));
+    }
+
+    @Override
+    public boolean isPaginationSupported() {
+        return false;
+    }
+
+    @Override
+    public boolean isRecordCountSupported() {
+        return true;
+    }
+
+    @Override
+    public void put(List<Record> records) throws AnalyticsException, AnalyticsTableNotAvailableException {
+        Collection<List<Record>> batches = GenericUtils.generateRecordBatches(records);
+        for (List<Record> batch : batches) {
+            addBatch(batch);
+        }
+    }
+
+    private void addBatch(List<Record> batch) throws AnalyticsException, AnalyticsTableNotAvailableException {
+        Record firstRecord = batch.get(0);
+        int tenantId = firstRecord.getTenantId();
+        String tableName = firstRecord.getTableName();
+        String collection = this.generateTargetCollectionName(tenantId, tableName);
+        try {
+            Map<String, Document> documentsToInsert = new HashMap<String, Document>();
+            for (Record record : batch) {
+                UpdateResult result = db.getCollection(collection).replaceOne((eq("_id", record.getId())),
```

Comment:
Here, still problem is, for each record in the batch, you're doing a database operation to update the one if it's already there. This would be slow when you need to write the record batches efficiently. So what you need to do is, first, try to insert the whole batch, with a "insertMany", only if it fails, then fallback to a record by record "replaceOne". You can actually probably check the performance difference after doing this change.